API Lesson 2 Part 1 flask updates (#146)
Ada-Developers-Academy/core-unit-2: api-2-read-404s/read-one-book.md

# Read One Book Endpoint

<!-- FLASK UPDATE -->

> <!-- <iframe src="https://adaacademy.hosted.panopto.com/Panopto/Pages/Embed.aspx?pid=781f275a-4b47-454a-a6e1-ae690178b4d2&autoplay=false&offerviewer=true&showtitle=true&showbrand=true&captions=true&interactivity=all" height="405" width="720" style="border: 1px solid #464646;" allowfullscreen allow="autoplay"></iframe> -->

Author: The video is hidden because the route decorator used is the old style.

For this feature, we should therefore make a `GET` request to the `/books` path, and include the `id` of the record to retrieve as part of the endpoint.

| HTTP Method | Endpoint   | Example   |
| ----------- | ---------- | ---------- |
| `GET`       | `/books/<book_id>` | `/books/1` |

### !callout-info

## Placeholder Syntax in Documentation Can Vary

Notice that in the endpoint example shown above, we used the syntax `<book_id>` to indicate a placeholder for the actual ID of the book that we want to retrieve. This syntax was chosen intentionally, as we'll see in just a moment when we implement the route. Another common method of seeing placeholders indicated in route documentation is with a colon, like `/books/:book_id`.

### !end-callout

`GET` requests do not include a request body, so no additional planning around the request body is needed.

> For the response, our feature asks that we send back information about a single book, made up of three pieces of data. We can express this as a single JSON object (roughly equivalent to a Python `dict`) with `id`, `title`, and `description` keys, sent in the response body. To indicate that we are successfully responding to a request for data, the response status should be `200 OK`.

Reviewer: This paragraph is a little dense, could we reformat the information a little to break it up? One possibility:

```suggestion
For the response, our feature asks that we send back information about a single book, made up of three pieces of data. We can express this as a single JSON object (roughly equivalent to a Python `dict`), which contains the keys:
- `id`
- `title`
- `description` 

Our JSON object is sent in the response body, and we indicate that we are successfully responding to a request for data by using the response status `200 OK`.
```

Our endpoint will need to:

1. Read the `book_id` from the request path
2. Retrieve the book object with the matching `book_id` from the "database" (currently our list of books)

> 3. Convert the book object so that it can be sent back through Flask with the appropriate structure (a single dictionary with `id`, `title`, and `description` keys)

Reviewer 2: Would it be confusing to mention that Flask will convert the dictionary to a JSON object? We mention that we send back a JSON object in the response body so I was thinking we could illustrate how that happens here but it could be too much all at once in the steps here?

Author: I went ahead and added a bit more to the next bullet (4). Let me know if this feels like it works to you.

Reviewer 2: Whoops, I missed your reply from Friday. 

Bullet 4 looks good in addition to the changes from bullets 1-3.

| `"/<book_id>"`                                      | This is the `Blueprint` path syntax to indicate _route parameters_. Flask won't literally look for `/<book_id>` in the path. Instead, it will treat the `<book_id>` as a placeholder that will receive the value of whatever follows `/books/`. As described in the note below, we'll need to be a little careful of how we use this value. Flask uses the name of the placeholder to pass the value into our route function _by name_ as a keyword argument. So the name we use here should be descriptive, and must match the parameter name we use to receive the value in our route function.                                                                                               |

> | `def get_one_book(`&#8203;`book_id):`                                | This `get_one_book` function is called whenever an HTTP request matches its route decorator. The name of the function should describe the purpose of the route, but otherwise has no real requirements. Because our route includes a route parameter, we must include a function parameter in this method to receive the matched value, and its name must be the same as the route parameter placeholder name. Since we used `book_id` as our placeholder, the function parameter must also be called `book_id`. |

Reviewer: What is the character `&#8203;` doing for formatting on this line?

Author: It's a zero-width space. It's allowing the `get_one_book(` to stay on the same line as the `def`, then only the `book_id):` wraps to the next line. Without it, the `def` was all alone (looked very odd), and I didn't want to try adjusting the table column widths.

> |`book_id = int(book_id)` | We must convert the `book_id` matched in the URL from a string to an int. Otherwise, it won't be possible for it to compare equal to any of our book's ID values, since they are stored as integers. URLs are strings, and any placeholder value will also be a string by default. There are ways to hint at the expected type for a placeholder that you may wish to investigate on your own. |

Reviewer 2: ```suggestion
|`book_id = int(book_id)` | We must convert the `book_id` matched in the URL from a string to an int. Otherwise, it won't be possible to compare if the incoming `book_id` is equal to any of our book's ID values, since they are stored as integers. URLs are strings, and any placeholder value will also be a string by default. There are ways to hint at the expected type for a placeholder that you may wish to investigate on your own. |
```

Author: Expanding out the "it" sounds like a good idea. I'd just do a minor revision to your suggestion.
```suggestion
|`book_id = int(book_id)` | We must convert the `book_id` matched in the URL from a string to an int. Otherwise, it won't be possible to check whether the incoming `book_id` is equal to any of our book's ID values, since they are stored as integers. URLs are strings, and any placeholder value will also be a string by default. There are ways to hint at the expected type for a placeholder that you may wish to investigate on your own. |
```

> |`book_id = int(book_id)` | We must convert the `book_id` matched in the URL from a string to an int. Otherwise, it won't be possible for it to compare equal to any of our book's ID values, since they are stored as integers. URLs are strings, and any placeholder value will also be a string by default. There are ways to hint at the expected type for a placeholder that you may wish to investigate on your own. |

Author: Resolving as a duplicate of the version that was included in your main feedback.

### !callout-warning

## Python Doesn't Know What a `book_id` Is

We named the route parameter `book_id` because we expect it to be an ID of a book. But just like regular variable names, Python has no idea what a `book_id` is. All it does is look for whatever part of the URL path follows `/books/` and captures that into the variable `book_id`. A request for `/books/360` would set the value of `book_id` to be `"360"`. A request for `/books/tacocat` would set the value of `book_id` to be `"tacocat"`!

<br />

We should be careful to avoid thinking that Python uses the name of the parameter to do any kind of validation. The name provides information to us developers, not to Python. Flask does provide a way to specify a [converter](https://flask.palletsprojects.com/en/3.0.x/api/#url-route-registrations) for route parameters that can help with validation, but we won't cover that in this curriculum.

### !end-callout

> ## An Imperfect Solution

Author: This section is added in response to this comment: https://app.asana.com/0/0/1206697443678567/1206923601980905/f
